dm/master/shardddl/pessimist.go

```diff
@@ -0,0 +1,376 @@
+// Copyright 2020 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package shardddl
+
+import (
+	"context"
+	"sync"
+
+	"go.etcd.io/etcd/clientv3"
+	"go.uber.org/zap"
+
+	"github.com/pingcap/dm/pkg/log"
+	"github.com/pingcap/dm/pkg/shardddl/pessimism"
+)
+
+// Pessimist used to coordinate the shard DDL migration in pessimism mode.
+type Pessimist struct {
+	mu sync.Mutex
+
+	logger log.Logger
+
+	closed bool
+	cancel context.CancelFunc
+	wg     sync.WaitGroup
+
+	cli *clientv3.Client
+	lk  *pessimism.LockKeeper
+
+	// sources used to get all sources relative to the give task.
+	sources func(task string) []string
```

Comment:
How about refract to `taskSources` or `getTaskSources`? `sources` seems weird.

Reply:
renamed in https://github.com/pingcap/dm/pull/456/commits/e2972a29a148ff0c49b84c63b4066e8f09459e0d.